File: Java/Lesson051_GenerateParentheses.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 051 -- Generate Parentheses
// Category   : Stack
// Difficulty : Medium
// Study Plan : Day 25
// =============================================================
//
// QUESTION:
//   Given n, return all valid combinations of n pairs of parentheses.
//
//   Example: n=3 -> ["((()))","(()())","(())()","()(())","()()()"]
// =============================================================

import java.util.*;
public class Lesson051_GenerateParentheses {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        bt(res, "", 0, 0, n);
        return res;
    }
    private void bt(List<String> r, String cur, int o, int c, int n) {
        if (cur.length()==2*n) { r.add(cur); return; }
        if (o<n) bt(r, cur+"(", o+1, c, n);
        if (c<o) bt(r, cur+")", o, c+1, n);
    }
    public static void main(String[] a){
        System.out.println(new Lesson051_GenerateParentheses().generateParenthesis(3));
    }
}

File: Go/Lesson_051_generate_parentheses.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 051 -- Generate Parentheses
// Category   : Stack
// Difficulty : Medium
// Study Plan : Day 25
// =============================================================
//
// QUESTION:
//   Given n, return all valid combinations of n pairs of parentheses.
//
//   Example: n=3 -> ["((()))","(()())","(())()","()(())","()()()"]
// =============================================================

package main
import "fmt"
func generateParenthesis(n int) []string {
    res := []string{}
    var bt func(cur string, o, c int)
    bt = func(cur string, o, c int) {
        if len(cur) == 2*n { res = append(res, cur); return }
        if o < n { bt(cur+"(", o+1, c) }
        if c < o { bt(cur+")", o, c+1) }
    }
    bt("", 0, 0)
    return res
}
func main(){ fmt.Println(generateParenthesis(3)) }
